src/Hamiltonians/Product.jl

```diff
@@ -0,0 +1,166 @@
+"""
+    HamiltonianProduct(A::AbstractHamiltonian, B::AbstractHamiltonian; commuting=A==B)
+
+The product of two [`AbstractHamiltonian`](@ref)s, acting from right to left. The two Hamiltonians
+must act on the same address space. Set `commuting` to `true` if `A` and `B` commute.
+"""
+struct HamiltonianProduct{T, O1<:AbstractHamiltonian, O2<:AbstractHamiltonian, C} <: AbstractHamiltonian{T}
+    op1::O1
+    op2::O2
+end
+function HamiltonianProduct(
+    op1::AbstractHamiltonian{T1}, op2::AbstractHamiltonian{T2}; commuting = op1==op2
```

Comment:
If either `op1` or `op2` diagonal (does it also need to be real?), it commutes as well, right? Should we add that or is it not necessary?